### PR TITLE
Eliminate possible NULL char in html which breaks nokogiri

### PR DESCRIPTION
**Pull request by tak1n**
We face a issue where the src attribute of an img tag is not found because the file contains a NULL char (it's fetched from a windows server).

The problematic html part looks like following in vim:
selection_002

This `^@` is a NULL Character in ASCII (http://superuser.com/a/75133)
This leads to following parsed html elements (through nokogiri):

```
[7] pry(main)> doc.css('[data-absatz="14AbsatzLi"]').first
=> #(Element:0x2ab7ac9e0be0 {
  name = "div",
  attributes = [ #(Attr:0x2ab7ac9df36c { name = "data-absatz", value = "14AbsatzLi" }), #(Attr:0x2ab7ac9df358 { name = "class", value = "itemposition picture" }), #(Attr:0x2ab7ac9df344 { name = "rel", value = "10235274" })],
  children = [
    #(Text "\r\n"),
    #(Element:0x2ab7ac9b92fc {
      name = "div",
      attributes = [
        #(Attr:0x2ab7ac9b88e8 { name = "itemprop", value = "image" }),
        #(Attr:0x2ab7ac9b88d4 { name = "itemscope", value = "" }),
        #(Attr:0x2ab7ac9b88c0 { name = "itemtype", value = "http://schema.org/ImageObject" }),
        #(Attr:0x2ab7ac9b88ac { name = "class", value = "bildbox inline_element inline grid_8" })],
      children = [
        #(Text "\r\n"),
        #(Element:0x2ab7ac9882d8 {
          name = "div",
          attributes = [ #(Attr:0x2ab7ac985948 { name = "class", value = "photowrap" })],
          children = [
            #(Text "\r\n"),
            #(Element:0x2ab7ac971560 { name = "div", attributes = [ #(Attr:0x2ab7ac970c00 { name = "class", value = "corner top_left" })] }),
            #(Text "\r\n"),
            #(Element:0x2ab7ac93e19c { name = "div", attributes = [ #(Attr:0x2ab7ac93978c { name = "class", value = "corner top_right" })] }),
            #(Text "\r\n"),
            #(Element:0x2ab7ac8fd9bc { name = "div", attributes = [ #(Attr:0x2ab7ac8fcd78 { name = "class", value = "corner bottom_left" })] }),
            #(Text "\r\n"),
            #(Element:0x2ab7ac8ef434 { name = "div", attributes = [ #(Attr:0x2ab7ac8ee82c { name = "class", value = "corner bottom_right" })] }),
            #(Text "\r\n"),
            #(Element:0x2ab7ac8c75c4 { name = "img", attributes = [ #(Attr:0x2ab7ac8c6534 { name = "itemprop", value = "url" }), #(Attr:0x2ab7ac8c6520 { name = "alt", value = "" })] })]
          })]
      })]
  })
```

We can clearly see the src attribute of the img tag is missing in the parsed object.
After some googling I found https://github.com/sparklemotion/nokogiri/issues/1240 where they suggest the gem user of nokogiri needs to substitute this NULL char with `""`. 

This bug causes a crash in https://github.com/LawrenceWoodman/mida where we pass in the already parsed nokogiri document (which is fetched by MetaInspector). Therefore it seems logical for me that MetaInspector handles this NULL char conversion after fetching the html.

CC @jcoyne not sure if you want to add this to https://github.com/LawrenceWoodman/mida/blob/master/lib/mida/document.rb#L19 for the case a user creates a new doc with given html or you delegate the responsibility of that to the user/lib which fetches the html from the server.

### CODE REVIEW

**Comment by koni84** [comment]
+1

**Comment by fionita** [comment]
+1

File: lib/meta_inspector/request.rb [hunk 1]
```diff
@@ -28,7 +28,7 @@ def initialize(initial_url, options = {})
     delegate :url => :@url
 
     def read
-      response.body if response
+      response.body.gsub(/\0/,"") if response
```
Comment (jcoyne) [inline lib/meta_inspector/request.rb]: I wonder if it would be better to use `String#tr` instead of `#gsub`

Comment (tak1n) [inline lib/meta_inspector/request.rb]: Yep good catch :)

According to https://github.com/JuanitoFatas/fast-ruby#stringgsub-vs-stringtr-code `String#tr` is faster than `String#gsub`.

Changing to `String#tr`

**Comment by jaimeiniesta** [comment]
Thanks a lot for your contribution, it looks good to me! :+1:

Can you please add a test for this change and squash the commits into one?

Take a look at [how we organize fake web responses](https://github.com/jaimeiniesta/metainspector/blob/master/spec/spec_helper.rb#L21-L27) to set up the smallest possible example, please.

**Comment by tak1n** [comment]
@jaimeiniesta done, I added it to encoding specs if thats ok?

**Comment by jaimeiniesta** [comment]
@tak1n thanks, the encoding specs is a good place for that. However, I just see this line added, where's the NULL char?

``

**Comment by jaimeiniesta** [comment]
It might be good to add an HTML comment on the line where the NULL char is for better understanding.

**Comment by tak1n** [comment]
@jaimeiniesta I think it depends on what editor (encoding?) u use, with vim I see it:

selection_002

Also with the default ubuntu editor (gedit) I can see it:

selection_002

Added the HTML comment in the response fixture file.

**Comment by jaimeiniesta** [comment]
Ah, I couldn't see it in Atom, but I see it using vim.

Can you add an HTML comment before that line, explaining that the alt attribute contains a NULL char? With that we're good to merge.

**Comment by tak1n** [comment]
@jaimeiniesta I already added the HTML comment.

`<!-- This image tag contains a NULL character in alt text, this is used to test https://github.com/sparklemotion/nokogiri/issues/1240, which is handled in MetaInspector::Request#read -->`

**Comment by jaimeiniesta** [comment]
Merged and released as 5.2.2.

Thanks!

**Comment by tak1n** [comment]
Thanks :smile:

**Pull request merged**